### PR TITLE
Implements integration test

### PR DESCRIPTION
**Pull request by LynxPlay101**
As pina-golada did struggle a lot in the latest releases, this commit
adds an integration test to the project which compiles a test project
and tests for correct compilation and write to disk functionality.

As pina-golada generation cannot be dynamically loaded/unloaded, the implemented integration tests basically consist of three states. 

1. The first tests that are run are the pre-tests which generate the asset implementation as well as set up the file permission on the file system. 
2. The second tests are run as an integration which tests for the actual pina-golada code working.
3. The third tests are run as the post-tests which delete local test files and cleans up the pina-golada source.

### CODE REVIEW

File: internal/golada/builder/builder.go [hunk 1]
```diff
@@ -165,6 +165,10 @@ func (b Builder) BuildFile() (by []byte, err error) {
 			return nil, e
 		}
 
+		for _, file := range topLevelDir.Files() {
```
Comment (LynxPlay101) [inline internal/golada/builder/builder.go]: This line here adds files that are in the top-level directory to the debug output

File: pkg/annotation/csv_parser.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func (c *CsvParser) Parse(comment string, annotation Annotation) (e error) {
 	keyValuePair := strings.Split(s, ";")
 	for _, paired := range keyValuePair {
 		keyToValue := strings.Split(paired, ",")
-		if len(keyValuePair) >= 2 {
+		if len(keyToValue) >= 2 {
```
Comment (LynxPlay101) [inline pkg/annotation/csv_parser.go]: This is actually a fun bug that was never noticed. It failed to parse annotations if the annotation only defines one key-value pair :/ both gonut and watchful still define package in the annotation tho which means this was never noticed.

File: internal/golada/cmd/cleanup.go [hunk 3]
```diff
@@ -45,37 +45,42 @@ var cleanupCommand = &cobra.Command{
 			l = logger.NewDefaultLogger(os.Stdout, logger.Info)
 		}
 
-		e := filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
```
Comment (LynxPlay101) [inline internal/golada/cmd/cleanup.go]: Extracts into its own function to be called by the integration test

File: pkg/files/directory.go [hunk 4]
```diff
@@ -306,6 +306,7 @@ func copyDirectory(original Directory, new Directory) error {
 // NewRootDirectory returns a new root directory
 func NewRootDirectory() Directory {
 	return &memoryDirectory{
-		name: paths.RootPath(),
+		name:     paths.RootPath(),
+		PermBits: os.FileMode(0755),
```
Comment (LynxPlay101) [inline pkg/files/directory.go]: This here may be something we have to debate over. @HeavyWombat the permission we set here will as it defines the permissions for the folders that are created without being backed by assets.

Eg: have an asset:
```yml
- golang
  - main.go
```

and you write it to `~/this/path/does/not/exist/yet`. Pina-Golada will with this change create those folders using 0755

Comment (HeavyWombat) [reply pkg/files/directory.go]: I think the best idea is to rely on the system's umask. I had no idea how Golang does it, so I had to read the `os` package documentation.

```go
package main

import "os"

func main() {
	os.Mkdir("foobar", 0777)
}
```

```
➜  ~ umask
022
➜  ~
➜  ~ go run main.go
➜  ~
➜  ~ ls -ldh foobar
drwxr-xr-x  2 mdiester  staff    64B May 21 22:14 foobar
➜  ~
```

As advertised in the doc, the `Mkdir` function works as expected with the umask. So setting `0777` (full rights) will create the desired `0755` as most systems have a default umask of `022`. In case we run on systems with another umask, we would honour the settings accordingly.

Long story short, I think it is fair to set `0777` in `Mkdir` and let the system take care of the rest.

File: pkg/annotation/csv_parser.go (hunk 2 above)
Comment (HeavyWombat) [reply pkg/annotation/csv_parser.go]: This might explain why I struggled to remove the package definition in `gonut` after I read that is not required anymore.

**Comment by HeavyWombat** [comment]
While you are at it, could you bump `gonvenience` to [v1.7.8](https://github.com/homeport/gonvenience/releases/tag/v1.7.8) or newer? There was a bug in the `bunt` package that *printf functions evaluated both the format string as well as the input arguments. In the case of `pina-golada`, this led to a weird output if filenames contains underscores as part of their filename.

**Comment by LynxPlay101** [comment]
> do

**Comment by LynxPlay101** [comment]
Yeah I have played around with the Before and AfterSuite features of ginkgo, but when I tried to implement those that way, I ran into some trouble. When generating the compiled test files, the go "class loader" or whatever it is called, cannot dynamically load the generated asset implementation. 

This ends up in the fact that the file exists when the suite starts, but the instance of it has not been injected into the Injector var in the interface file. I have been searching for a good way to dynamically load a file into the go executable, but this seems to be quite the impossible.

Therefore I ended up with three separate steps, recompiling the code after the generation is done to ensure proper asset instance injection

File: pkg/files/directory.go (hunk 4 above)
Comment (LynxPlay101) [reply pkg/files/directory.go]: Alright, we will default to 0777 permission and let the file system take care of umasking those permissions

**Comment by HeavyWombat** [comment]
There is a way, but it means that we have to break the paradigm that everything is done in one test binary run. What I mean is that the `BeforeSuite` function calls OS calls to create a sample application, uses the `go` command to run a developer version of `pina-golada` to create the asset source file and eventually compiles a binary for the test. This binary is later removed in the `AfterSuite` function. For an integration test I find it absolutely okay to do mimic how another project would use `pina-golada`.

**Comment by LynxPlay101** [comment]
So you would have a unit test that runs a pina-golada version on an integration project, then calls the unit tests in those projects and then cleans the assets again ?

**Comment by HeavyWombat** [comment]
Yes, I think this is what I meant. Or, as pseudo-code:
```go
package integration_test

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var _ = BeforeSuite(func() {
	err := createSampleApplicationRequiringPinaGoload("../assets/sample/main.go")
	Expect(err).ToNot(HaveOccurred())

	cmd := exec.Command("go", "run", "cmd/golada/main.go", "generate")
	Expect(cmd.Run()).ToNot(HaveOccurred())

	Expect(len(findGeneratedFiles())).To(BeEquivalent(1))
})

var _ = Describe("integration test", func() {
	Context("pgl exploiter", func() {
		It("should do things", func() {
			cmd := exec.Command("go", "run", "integration/test/main.go", "test-1")
			Expect(cmd.Run()).ToNot(HaveOccurred())

			ok, err := checkThings()
			Expect(err).ToNot(HaveOccurred())
			Expect(ok).To(BeTrue())
		})

		It("should do other things", func() {
			cmd := exec.Command("go", "run", "integration/test/main.go", "test-2")
			Expect(cmd.Run()).ToNot(HaveOccurred())

			ok, err := checkOtherThings()
			Expect(err).ToNot(HaveOccurred())
			Expect(ok).To(BeTrue())
		})
	})
})

var _ = AfterSuite(func() {
	cmd := exec.Command("go", "run", "cmd/golada/main.go", "cleanup")
	Expect(cmd.Run()).ToNot(HaveOccurred())

	Expect(len(findGeneratedFiles())).To(BeEquivalent(0))
})
```

**Comment by LynxPlay101** [comment]
Yeah makes sense. I do not believe the pina-golada unit test needs to call pina-golada to generate from the cli but the exec command makes sense. The question is, wouldn't it make sense to not call go but go test on the compiled directory? the go run approach seems like we would lose out on the ginkgo framework

**Comment by LynxPlay101** [comment]
@HeavyWombat bump ^

**Comment by LynxPlay101** [comment]
> I ran it locally to see how it looks from the `ginkgo` output and I like it. The only thing where I do not have a good idea is the directory structure. It feels a bit asymmetric to the rest.

Mind elaborating ? I mean the name ".test" folder name is preventing the ginkgo call on the pina-golada root to pick up on the integration tests before the test generation suits have been run.

**Comment by HeavyWombat** [comment]
The `.test` is fine. I meant that up until now, all tests (we only had unit tests) are colocated with the actual package code. The integration test code has its own root directory in the project directory. This is what I meant with asymmetry. However, I do not see an alternative to this at the moment. Maybe one thing could be to move the `integration` directory into a `test` directory to put it into the right focus.

**Comment by LynxPlay101** [comment]
Well as of right now, I am happy with those changes. If we have an idea on how to further improve on this, I believe we can move this into a future issue. Thanks for all the feedback 👍
